packages/backend/src/core/discovery/engine/DiscoveryEngine.ts

```diff
@@ -0,0 +1,40 @@
+import { AddressAnalyzer, Analysis } from '../analysis/AddressAnalyzer'
+import { DiscoveryConfig } from '../config/DiscoveryConfig'
+import { DiscoveryLogger } from '../utils/DiscoveryLogger'
+import { DiscoveryStack } from './DiscoveryStack'
+import { shouldSkip } from './shouldSkip'
+
+export class DiscoveryEngine {
+  constructor(
+    private readonly addressAnalyzer: AddressAnalyzer,
+    private readonly logger: DiscoveryLogger,
+  ) {}
+
+  async discover(config: DiscoveryConfig) {
+    const resolved: Analysis[] = []
+    const stack = new DiscoveryStack()
+    stack.push(config.initialAddresses, 0)
+
+    while (!stack.isEmpty()) {
+      const item = stack.pop()
+
+      const reason = shouldSkip(item, config)
```

Comment:
can't we put it as a method in DiscoveryStack?

Reply:
It's a pure function, I don't see a particular need.